tsdb/iterator.go

```diff
@@ -0,0 +1,120 @@
+package tsdb
+
+import (
+	"time"
+)
+
+// maxTime is used as the maximum time value when computing an unbounded range.
+var maxTime = time.Date(3000, 1, 1, 0, 0, 0, 0, time.UTC)
+
+type Iterator interface {
+}
+
+type FloatIterator interface {
+	Iterator
+	Next() *FloatValue
+}
+
+type FloatValue struct {
+	Time  time.Time
+	Value float64
+	Tags  map[string]string
+}
+
+// FloatMinIterator emits a float value for each interval.
+type FloatMinIterator struct {
+	// buffered input iterators
+	inputs bufFloatIterators
+
+	// Calculates the window based on offset from the epoch and window duration.
+	Offset   time.Duration
+	Interval time.Duration
+}
+
+// NewFloatMinIterator returns a new instance of
+func NewFloatMinIterator(inputs []FloatIterator) *FloatMinIterator {
+	return &FloatMinIterator{
+		inputs: newBufFloatIterators(inputs),
+	}
+}
+
+// Next returns the minimum value for the next available interval.
+func (itr *FloatMinIterator) Next() *FloatValue {
+	// Calculate next window.
+	_, endTime := itr.inputs.nextWindow(itr.Interval, itr.Offset)
+
+	// Consume all points from all iterators within the window.
+	var min *FloatValue
+	for _, input := range itr.inputs {
+		for {
+			v := input.Next()
+			if v == nil {
+				break
+			} else if !v.Time.Before(endTime) {
+				input.unread(v)
```

Comment:
Thanks. There should really be a `NextBefore(t time.Time)` on the buffered iterators. It would help clean up the min iterator.